torch/_subclasses/functional_tensor.py

```diff
@@ -0,0 +1,259 @@
+            # This also alleviates us from having to manually handle metadata mutations on the wrapper.
+            return func(*args_unwrapped, **kwargs)
+        # Originally I tried to implement my subclass without giving it a torch_dispatch, but I gave up:
+        # - _make_wrapper_subclass requires a __torch_dispatch__
+        # - If we want to use _make_subclass(), we have a problem: the subclass will share a TensorImpl with the inner tensor,
+        #   which is of type FunctionalTensorWrapper! We explicitly do not want our wrapper to be a FunctionalTensorWrapper.
+        # - If we use the default tensor.__new__(), we have another problem: it returns inner_tensor.alias(),
+        #   which causes every subclass created above autograd to have autograd view metadata (in addition to also being a FunctionalTensorWrapper).
+        raise RuntimeError(
+            "Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()"
+        )
+
+    def __repr__(self):
+        return f"FunctionalTensor({str(self.elem)})"
+
+    @staticmethod
+    def to_functional(x):
+        # We will do the wrapping for the user.
+        assert not torch._is_functional_tensor(x)
+        # The only autograd metadata we care about on the FunctionalTensor is:
+        # - requires_grad (so autograd runs)
+        # - is_leaf (so that mutations on graph inputs that are not leaves are allowed by the autograd engine)
+        #   this is handled by FunctionalTensor.to_functional
+        x_functional = torch._to_functional_tensor(x, mirror_autograd_meta=True)
+
+        torch._functionalize_enable_reapply_views(True)
+        if x.requires_grad and not x.is_leaf:
+            out = _ToFunctionalTensor.apply(x_functional)
+            return out
+        else:
+            out = FunctionalTensor(x_functional)
+            out.requires_grad = x_functional.requires_grad
+            return out
+
+    def from_functional(self):
+        torch._sync(self)
+        return torch._from_functional_tensor(self.elem)
+
+
+class FunctionalTensorMode(TorchDispatchMode):
+    def __init__(self):
+        self.is_active = True
+        self.is_on_stack = False
+        self.enter_stack = []
+        # Indicates to our torch_dispatch dispatching infra that
+        # this is an "infra" mode with lower dispatching precedence.
+        self._mode_key = torch._C._TorchDispatchModeKey.FUNCTIONAL
+
+    # No-op if FunctionalTensorMode is already in use
+    def __enter__(self):
+        if (
+            torch._C._get_dispatch_mode(torch._C._TorchDispatchModeKey.FUNCTIONAL)
+            is None
+        ):
+            self.enter_stack.append(True)
+            return super().__enter__()
+        else:
+            self.enter_stack.append(False)
+            return self
+
+    def __exit__(self, a, b, c):
+        is_on_stack = self.enter_stack.pop()
+        if is_on_stack:
+            super().__exit__(a, b, c)
```

Comment:
The adjusted semantics I suggested for FakeTensorMode should apply here; doubly so because there are far fewer preexisting uses of FunctionalTensorMode here